<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Контекст-исследования" data-toc-modified-id="Контекст-исследования-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Контекст исследования</a></span></li><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица books</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица publishers</a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица ratings</a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li></ul></li><li><span><a href="#Задание" data-toc-modified-id="Задание-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задание</a></span><ul class="toc-item"><li><span><a href="#Cколько-книг-вышло-после-1-января-2000-года?" data-toc-modified-id="Cколько-книг-вышло-после-1-января-2000-года?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Cколько книг вышло после 1 января 2000 года?</a></span></li><li><span><a href="#Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитаем количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-(исключим-из-анализа-брошюры)" data-toc-modified-id="Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-(исключим-из-анализа-брошюры)-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определим издательство, которое выпустило наибольшее число книг толще 50 страниц (исключим из анализа брошюры)</a></span></li><li><span><a href="#Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитываются-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитываются-только-книги-с-50-и-более-оценками-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определим автора с самой высокой средней оценкой книг — учитываются только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Результаты:" data-toc-modified-id="Результаты:-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Результаты:</a></span></li></ul></div>

## Контекст исследования

"Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.    
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта."

## Загрузка данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
#Чтобы выполнить SQL-запрос, используйте Pandas:
#pd.io.sql.read_sql(query, con = engine) 

### Таблица books

* book_id — идентификатор книги;
* author_id — идентификатор автора;
* title — название книги;
* num_pages — количество страниц;
* publication_date — дата публикации книги;
* publisher_id — идентификатор издателя.

In [3]:
query = '''SELECT * FROM BOOKS'''

In [4]:
books = pd.io.sql.read_sql(query, con = engine) 
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


### Таблица authors

* author_id — идентификатор автора;
* author — имя автора.

In [6]:
query = '''SELECT * FROM AUTHORS'''

In [7]:
authors = pd.io.sql.read_sql(query, con = engine) 
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


### Таблица publishers

* publisher_id — идентификатор издательства;
* publisher — название издательства;

In [9]:
query = '''SELECT * FROM PUBLISHERS'''

In [10]:
publishers = pd.io.sql.read_sql(query, con = engine) 
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [11]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


### Таблица ratings

* rating_id — идентификатор оценки;
* book_id — идентификатор книги;
* username — имя пользователя, оставившего оценку;
* rating — оценка книги.

In [12]:
query = '''SELECT * FROM RATINGS'''

In [13]:
ratings = pd.io.sql.read_sql(query, con = engine) 
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [14]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


### Таблица reviews

* review_id — идентификатор обзора;
* book_id — идентификатор книги;
* username — имя автора обзора;
* text — текст обзора.

In [15]:
query = '''SELECT * FROM REVIEWS'''

In [16]:
reviews = pd.io.sql.read_sql(query, con = engine) 
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [17]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


## Задание

### Cколько книг вышло после 1 января 2000 года?

In [18]:
query = '''
SELECT COUNT(DISTINCT TITLE) FROM BOOKS
WHERE PUBLICATION_DATE > '2000-01-01'
'''
pd.io.sql.read_sql(query, con = engine) 

count
0    818

### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [28]:
#query = '''
#SELECT B.BOOK_ID, TITLE, COUNT(DISTINCT R.REVIEW_ID) REVIEWS, ROUND(AVG(RATING), 2) RATING FROM BOOKS B
#LEFT JOIN REVIEWS R ON R.BOOK_ID = B.BOOK_ID
#LEFT JOIN RATINGS ON RATINGS.BOOK_ID = B.BOOK_ID
#GROUP BY B.BOOK_ID
#'''
#pd.io.sql.read_sql(query, con = engine)

In [20]:
query = '''
SELECT B.BOOK_ID, TITLE, COUNT(REVIEW_ID) REVIEWS, ROUND((SELECT AVG(RATING) FROM RATINGS WHERE RATINGS.BOOK_ID = B.BOOK_ID), 2) RATING FROM  BOOKS B, REVIEWS R
WHERE R.BOOK_ID = B.BOOK_ID
GROUP BY B.BOOK_ID
'''
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  reviews  \
0        652          The Body in the Library (Miss Marple  #3)        2   
1        273                                          Galápagos        2   
2         51                           A Tree Grows in Brooklyn        5   
3        951  Undaunted Courage: The Pioneering First Missio...        2   
4        839                                        The Prophet        4   
..       ...                                                ...      ...   
989       64                                Alice in Wonderland        4   
990       55          A Woman of Substance (Emma Harte Saga #1)        2   
991      148                                          Christine        3   
992      790  The Magicians' Guild (Black Magician Trilogy  #1)        2   
993      828                           The Plot Against America        2   

     rating  
0      4.50  
1      4.50  
2      4.25  
3      4.00  
4      4.29  
..      ...  
989    4.23  
990    5.00  
991    3.43  
992    3.50  
993    3.00  

[994 rows x 4 columns]

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц (исключим из анализа брошюры)

In [21]:
query = '''
SELECT PUBLISHER, COUNT(BOOK_ID) BOOKS FROM PUBLISHERS P
LEFT JOIN BOOKS B ON B.PUBLISHER_ID = P.PUBLISHER_ID
WHERE NUM_PAGES > 50
GROUP BY PUBLISHER
ORDER BY BOOKS DESC
LIMIT 1
'''
pd.io.sql.read_sql(query, con = engine)

publisher  books
0  Penguin Books     42

### Определим автора с самой высокой средней оценкой книг — учитываются только книги с 50 и более оценками

In [22]:
query = '''
SELECT A.AUTHOR_ID, A.AUTHOR, AVG(RATING) AVG_RATING FROM AUTHORS A 
JOIN BOOKS B ON B.AUTHOR_ID = A.AUTHOR_ID
JOIN RATINGS R ON R.BOOK_ID = B.BOOK_ID
WHERE B.BOOK_ID IN (SELECT BOOK_ID FROM RATINGS 
                    GROUP BY BOOK_ID
                    HAVING COUNT(RATING) >= 50)
GROUP BY A.AUTHOR_ID
ORDER BY AVG_RATING DESC

'''
pd.io.sql.read_sql(query, con = engine)

author_id                                             author  avg_rating
0         236                         J.K. Rowling/Mary GrandPré    4.287097
1         402                  Markus Zusak/Cao Xuân Việt Khương    4.264151
2         240                                     J.R.R. Tolkien    4.246914
3         376                                  Louisa May Alcott    4.192308
4         498                                       Rick Riordan    4.080645
5         621                                    William Golding    3.901408
6         235                                      J.D. Salinger    3.825581
7         469           Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8         630  William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9         372                                         Lois Lowry    3.750000
10        106                                          Dan Brown    3.741259
11        195            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12        554                                    Stephenie Meyer    3.662500
13        311                                     John Steinbeck    3.622951

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [23]:
query = '''
SELECT AVG(REVIEWS) FROM
(SELECT COUNT(REVIEW_ID) REVIEWS FROM REVIEWS
WHERE USERNAME IN (SELECT USERNAME FROM RATINGS
                    GROUP BY USERNAME
                    HAVING COUNT(RATINGS) > 50)
GROUP BY USERNAME) REVIEWS_COUNT
  
'''
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

## Результаты:
* после 1 января 2000 года вышло 818 книг
* издательство Penguin Books выпустило наибольшее количество книг (объемом более 50 страниц)
* J.K. Rowling автор с самым высоким средним рейтингом книг, которые получили более 50 оценок
* Пользователи, которые поставили более 50 оценок, в среднем оставляют 24 отзыва